termui/label_select.go

```diff
@@ -0,0 +1,245 @@
+}
+
+func (ls *labelSelect) SetBug(cache *cache.RepoCache, bug *cache.BugCache) {
+	ls.cache = cache
+	ls.bug = bug
+	ls.labels = cache.ValidLabels()
+
+	// Find which labels are currently applied to the bug
+	bugLabels := bug.Snapshot().Labels
+	labelSelect := make([]bool, len(ls.labels))
+	for i, label := range ls.labels {
+		for _, bugLabel := range bugLabels {
+			if label == bugLabel {
+				labelSelect[i] = true
+				break
+			}
+		}
+	}
+
+	ls.labelSelect = labelSelect
+	ls.selected = 0
+}
+
+func (ls *labelSelect) keybindings(g *gocui.Gui) error {
+	// Abort
+	if err := g.SetKeybinding(labelSelectView, gocui.KeyEsc, gocui.ModNone, ls.abort); err != nil {
+		return err
+	}
+	// Save and return
+	if err := g.SetKeybinding(labelSelectView, 'q', gocui.ModNone, ls.saveAndReturn); err != nil {
+		return err
+	}
+	// Up
+	if err := g.SetKeybinding(labelSelectView, gocui.KeyArrowUp, gocui.ModNone, ls.selectPrevious); err != nil {
+		return err
+	}
+	if err := g.SetKeybinding(labelSelectView, 'k', gocui.ModNone, ls.selectPrevious); err != nil {
+		return err
+	}
+	// Down
+	if err := g.SetKeybinding(labelSelectView, gocui.KeyArrowDown, gocui.ModNone, ls.selectNext); err != nil {
+		return err
+	}
+	if err := g.SetKeybinding(labelSelectView, 'j', gocui.ModNone, ls.selectNext); err != nil {
+		return err
+	}
+	// Select
+	if err := g.SetKeybinding(labelSelectView, gocui.KeySpace, gocui.ModNone, ls.selectItem); err != nil {
+		return err
+	}
+	if err := g.SetKeybinding(labelSelectView, 'x', gocui.ModNone, ls.selectItem); err != nil {
+		return err
+	}
+	if err := g.SetKeybinding(labelSelectView, gocui.KeyEnter, gocui.ModNone, ls.selectItem); err != nil {
+		return err
+	}
+	// Add
+	if err := g.SetKeybinding(labelSelectView, 'a', gocui.ModNone, ls.addItem); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (ls *labelSelect) layout(g *gocui.Gui) error {
+	maxX, maxY := g.Size()
+	ls.childViews = nil
+
+	// TODO: Make width adaptive
+	width := 30
+	height := 2*len(ls.labels) + 3
+	x0 := 2
+	y0 := 2
+
+	v, err := g.SetView(labelSelectView, x0, y0, x0+width, y0+height)
+	if err != nil {
+		if err != gocui.ErrUnknownView {
+			return err
+		}
+	
+		v.Frame = false
+	}
+	y0 += 1
+
+	for i, label := range ls.labels {
+		viewname := fmt.Sprintf("view%d", i)
+		v, err := g.SetView(viewname, x0+2, y0, x0+width-2, y0+2)
+		if err != nil && err != gocui.ErrUnknownView {
+			return err
+		}
+		ls.childViews = append(ls.childViews, viewname)
+		v.Frame = i == ls.selected
+		v.Clear()
+		selectBox := " [ ] "
+		if ls.labelSelect[i] {
+			selectBox = " [x] "
+		}
+		fmt.Fprint(v, selectBox, label)
+		y0 += 2
+	}
+
+	v, err = g.SetView(labelSelectInstructionsView, -1, maxY-2, maxX, maxY)
+	ls.childViews = append(ls.childViews, labelSelectInstructionsView)
+	if err != nil {
+		if err != gocui.ErrUnknownView{
+			return err
+		}
+		v.Frame = false
+		v.BgColor = gocui.ColorBlue
+	}
+	v.Clear()
+	fmt.Fprint(v, "[↓↑,jk] Nav [a] Add item [q] Save and close")
```

Comment:
to be consistent with the other windows, the order should be quit / navigation / actions